<h1>Post</h1>

<p>It's been a few weeks since I blogged about <a href="https://github.com/ibm-functions/composer">IBM Composer</a>, sorry about that, flying to China and getting a kid will put a kink into your blogging schedule. ;) Today I want to share a simple demo of how to wrap serverless functions with try/catch and try/catch/finally logic. Let's start off with a simple function.</p>
<pre><code class="language-javascript">
function main(args) {

    if(!args.input) args.input = 0;

    if(args.input === 0) {
        throw new Error("Can't divide by zero and maintain the Universe.");
    }

    return { result: 10/args.input };

}
</code></pre>
<p>This function simply takes an input value and then returns the result of dividing it into ten. If the input is 0, an error is thrown. I called this function <code>tendividedby</code>. Here's a sample result using an input of 2:</p>
<pre><code class="language-javascript">{
    "result": 5
}
</code></pre>
<p>And here is a result when the input is 0:</p>
<pre><code class="language-javascript">{
    "error": "An error has occurred: Error: Can't divide by zero and maintain the Universe."
}
</code></pre>
<p>If you get the activation record for the last test, you can clearly see it reported as an error:</p>
<pre><code class="language-javascript">"response": {
  "status": "action developer error",
  "statusCode": 0,
  "success": false,
  "result": {
    "error": "An error has occurred: Error: Can't divide by zero and maintain the Universe."
  }
},
</code></pre>
<p>Alright, so what if we <em>don't</em> want an error reported? This is where a try/catch composition will come in handy. I begin by creating a new file for my composition, app.js. Here's the code.</p>
<pre><code class="language-javascript">composer.try(
    'safeToDelete/tendividedby',
    args => ({result:'invalid input'})
);
</code></pre>
<p>The composition consists of one command, <code>composer.try</code>. This command takes two arguments - the action to try running and an action to run on failure. In the example above an inline action is being used but you can definitely pass the name of an existing action instead. My inline action simply says to return a result with a string indicating the error. You could do other things of course, for example sending an email about the error so people could be notified.</p>
<p>So that's try/catch, nice and simple. How about try/catch/finally? While this isn't built into the composer function itself, you can &quot;fake&quot; it by simply using a sequence where the &quot;finally&quot; part comes after the try/catch. Here's how that could look:</p>
<pre><code class="language-javascript">composer.sequence(
    composer.try(
        'safeToDelete/tendividedby',
        args => ({result:'invalid input'})
    ),
    'safeToDelete/final'        
)
</code></pre>
<p>In this case, I've set the &quot;finally&quot; part of my logic to be an action called final. All that does is add a timestamp:</p>
<pre><code class="language-javascript">function main(args) {
    //just take the result and add a time
    args.timestamp = new Date();

    return args;
}
</code></pre>
<p>In case you're curious, this could also be an inline function, but I wanted to demonstrate a mixture of both. To start working with this composition, first I create it:</p>
<pre><code>fsh app create trycatchdemo app.js
</code></pre>
<p>Then I invoke it:</p>
<pre><code>fsh app invoke trycatchdemo --param input 5
</code></pre>
<p>This results in:</p>
<pre><code class="language-javascript">{
  result: 2,
  timestamp: "2017-11-22T13:52:32.177Z"
}
</code></pre>
<p>And here is the error condition version:</p>
<pre><code class="language-javascript">{
  result: "invalid input",
  timestamp: "2017-11-22T13:52:53.489Z"
}
</code></pre>
<p>What's cool is that the Composer's graphical shell does a kick butt job of rendering these calls. So for example, here is the good test:</p>
<p><img src="https://static.raymondcamden.com/images/2017/11/trycatchgood2.jpg" alt="Good example"></p>
<p>And here is the bad run:</p>
<p><img src="https://static.raymondcamden.com/images/2017/11/trycatchbad2a.jpg" alt="Bad example"></p>
<p>If you have any questions about these examples, let me know!</p>
}